src/equisolve/module.py

```diff
@@ -1,49 +1,132 @@
-TModule = TypeVar("TModule", bound="Module")
+
+class NumpyModule(metaclass=ABCMeta):
+    @abstractmethod
+    def forward(self, *args, **kwargs):
+        return
+
+    def __call__(self, *args, **kwargs):
+        return self.forward(*args, **kwargs)
+
+    #def __setattr__(self, name, value):
+    #    # TODO used to keep track of the parameters that
+    #           have been set up
+    #    # https://stackoverflow.com/a/54994818
+    #    return
+
+    def state_dict(self) -> OrderedDict:
+        """
+        All required parameteres to initialize a fitted module
+        """
+
+        # PR COMMENT: not implemented, but should be analogous
+        #             to torch.nn.Module by using __setattr_
+        return
+
+    # PR comment: Torch does not return a boolean but a
+    #             torch.nn.modules.module._IncompatibleKeys
+    #             need to think about this, if we keep bool
+    #             or make something analogous
+    def load_state_dict(self, state_dict: OrderedDict) -> bool:
+        """
+        Initialize a fitted module
+        """
+        # PR COMMENT: not implemented, but should be analogous
+        #             to torch.nn.Module by using __setattr_
+        return
 
 try:
     import torch
+    HAS_TORCH = True
     Module = torch.nn.Module
 except:
-    class Module(metaclass=ABCMeta):
-        @abstractmethod
-        def forward(self, *args, **kwargs):
-            return
+    HAS_TORCH = False
+    Module = NumpyModule
 
 
-TEstimatorModule = TypeVar("TEstimatorModule", bound="EstimatorModule")
+# PR COMMENT:
+# We want to allow people to switch to a torch model if needed
+# (to use the TorchScript for low level integration into MD code).
+# For that we need both class definitions (numpy and torch)
+# so we can convert a NumpyModule to a torch Module
+# the use case I see is that you did not work with torch
+# then you have a class based on NumpyModule, but you need
+# a class based on TorchModule, so we provide an utility function
+# that takes all the parameters (state_dict) of the numpy module,
+# converts them to torch tensors, and reinitialize it with 
+# the corresponding torch module
 
+# all modules that have a corresponding torch module
+# are stored here
+NUMPY_TO_TORCH_MODULE = {}
 
-class EstimatorModule(Module, metaclass=ABCMeta):
-    def forward(self, X: TensorMap):
-        return self.predict(X)
+# PR COMMENT I think this cannot be hiden a class decorator like
 
-    @abstractmethod
-    def fit(self, X: TensorMap, y: TensorMap) -> TEstimatorModule:
-        return
+#              # automatically creates EstimatorModuleNumpy and EstimatorModuleTorch 
+#              @equisolve.module
+#              class EstimatorModule(Module)
+#                  ...
 
-    @abstractmethod
-    def predict(self, X: TensorMap) -> TensorMap:
-        return
+#            because we would need to change the base class
+#            and that is super hacky https://stackoverflow.com/a/9541560
```

Comment:
A possible solution (taking inspiration from rust [proc-macros](https://doc.rust-lang.org/reference/procedural-macros.html)) would be to have the decorator parse the code and modify the corresponding AST/Token tree. I.e. instead of doing this dynamically at runtime like in the stackoverflow, do it at "compile" time

Feels like this should be possible (but I did not dig into it at all) with 

```
@equisolve.module
class EstimatorModule:
```

Reply:
So PEP638 looks exactly what I was thinking about, so it is out of question for now. Taking a step back, why are we trying to do this? Defining the base to be `torch.nn.Module` only if torch is available does not seems to have that much drawbacks to me.

Reply:
The approach

```python
class NumpyModule:
    # magic to work like a torch.nn.Module

class _RidgeBase: 
    pass
    # implementation of Ridge

class NumpyRidge(NumpyModule, _RidgeBase): pass

class TorchRidge(TorchModule, _RidgeBase): pass
```
seems to work. I tried this script which checks different ways of inheritance (B2 would be our approach)

```python
import torch

class A: # inheritance from object works with jit script
    def __init__(self, a : int):
        self.a = a

    def __call__(self, a :int):
        self.forward(a)

    def forward(self, a : int):
        return a

class _B:
    def __init__(self, a : int):
        self.a = a

    def forward(self, a : int):
        print(a)

class B(_B): # does not work because we inherit from a non-object class _B
    def __init__(self, a : int):
        _B.__init__(self, a)

    def __call__(self, a :int):
        self.forward(a)

class B2(torch.nn.Module, _B): # works because we still inherit from torch.nn.Module
    def __init__(self, a : int):
        _B.__init__(self, a)
        torch.nn.Module.__init__(self)

    def forward(self, a : int):
        return a

aobj = A(1)
torch.jit.script(aobj)
print('aobj compiled')
try:
    sm = torch.jit.trace(aobj, (torch.tensor(1),))
    print('aobj traced')
except RuntimeError as e:
    print('aobj not traced, error:')
    print('  RuntimeError:', str(e).replace('\n', '\n  '))

bobj = B(1)
try:
    torch.jit.script(bobj)
    print('bobj compiled')
except RuntimeError as e:
    print('bobj not compiled, error:')
    print('  RuntimeError:', str(e).replace('\n', '\n  '))

b2obj = B2(1)
torch.jit.script(b2obj)
sm = torch.jit.trace(b2obj, (torch.tensor(1),))
sm.eval()
torch.jit.freeze(sm)
print('b2obj run completely')
```

With output
```
aobj compile
aobj not traced, error:
  RuntimeError: Could not get name of python class object
bobj not compiled, error:
  RuntimeError: 
  Tried to set an attribute: a on a non-class: Tensor (inferred):
    File "/home/alexgo/code/script_trace/inheritance/inheritance.py", line 15
      def __init__(self, a : int):
          self.a = a
          ~~~~~~~~~~ <--- HERE
  '__init__' is being compiled since it was called from 'B.__init__'
    File "/home/alexgo/code/script_trace/inheritance/inheritance.py", line 22
      def __init__(self, a : int):
          _B.__init__(self, a)
          ~~~~~~~~~~~~~~~~~~~ <--- HERE
  
b2obj run completely

```

Reply:
Option B2 it is then!